man/uncrustify.1
.TH UNCRUSTIFY "1" "Oct 2009" "uncrustify 0.56" "User Commands"
.SH NAME
uncrustify \- C, C++, C#, D, Java and Pawn source code beautifier

.SH SYNOPSIS
.B uncrustify \fR[\fIOPTIONS\fR] [\fIFILES\fR]

.SH DESCRIPTION

If no input files are specified, the input is read from stdin.
.br
If reading from stdin, you should specify the language using \-l.

If \-F is used or files are specified on the command line, the output filename is
\fIPFX\fR + "/" + filename + \fISFX\fR.
.br
Unless, of course, the options \fB\-\-replace\fR or \fB\-\-no\-backup\fR are used.

When reading from stdin or doing a single file via the '\-f' option,
the output is dumped to stdout, unless redirected with \-o FILE.

Errors are always dumped to stderr

.SH OPTIONS
.SS "Basic Options:"
.TP
\fB\-c\fI CFG
Use the config file \fICFG\fR.
.br
If not specified, uncrustify will use \fB$UNCRUSTIFY_CONFIG\fR or \fB$HOME\\.uncrustify.cfg\fR.
.TP
\fB\-f\fI FILE
Process the single file \fIFILE\fR, sending output to stdout or the file specified with \fB\-o\fR.
.TP
\fB\-o\fI FILE
Redirect output to \fIFILE\fR.
.br
Use with \fB\-f\fR, \fB\-\-update\-config\fR, \fB\-\-update\-config\-with\-doc\fR,
\fB\-\-universalindent\fR.
.TP
\fB\-F\fI FILE
Read files to process from \fIFILE\fR, one filename per line.
.br
You can create this file using something like \'\fBfind . \-name "*.c" > list.txt\fR\'.
.br
This cannot be combined with \fB\-f\fR.
.TP
\fB\-\-prefix\fI PFX
Prepend \fIPFX\fR to the output filename path.
.br
This cannot be combined with \fB\-f\fR, \fB\-\-replace\fR, or \fB\-\-no\-backup\fR.
.TP
\fB\-\-suffix\fI SFX
Append \fISFX\fR to the output filename.
.br
The default is '.uncrustify' if neither \fISFX\fR or \fIPFX\fR are specified.
.br
This cannot be combined with \fB\-f\fR, \fB\-\-replace\fR, or \fB\-\-no\-backup\fR.
.TP
\fB\-\-replace\fR
Replace source files (creates a backup).
.br
This cannot be combined with \fB\-f\fR, \fB\-\-prefix\fR, or \fB\-\-suffix\fR.
.TP
\fB\-\-no\-backup\fR
Replace files, no backup. Useful if files are under source control
.br
This cannot be combined with \fB\-f\fR, \fB\-\-prefix\fR, or \fB\-\-suffix\fR.
.TP
\fB\-\-mtime\fR
Preserve mtime on replaced files.
.TP
\fB\-l\fR
Language override: C, CPP, D, CS, JAVA, PAWN, VALA, OC, OC+
.TP
\fB\-t\fR
Load a file with types (usually not needed)
.TP
\fB\-q\fR
Quiet mode \- no output on stderr (\fB\-L\fR will override)
.SS "Config/Help Options:"
.TP
\fB\-h\fR \-? \fB\-\-help\fR \fB\-\-usage\fR
Print this message and exit
.TP
\fB\-\-version\fR
Print the version and exit
.TP
\fB\-\-show\-config\fR
Print out option documentation and exit
.TP
\fB\-\-update\-config\fR
Output a new config file.
.TP
\fB\-\-update\-config\-with\-doc\fR
Output a new config file with embedded usage comments.
.TP
\fB\-\-universalindent\fR
Output a config file for Universal Indent GUI.
.TP
\fB\-\-detect\fR
Detects the config from a source file. Use with '\-f FILE'.
Detection is currently fairly limited.

.SS "Debug Options:"
.TP
\fB\-p\fI FILE
Dump debug info to a file
.TP
\fB\-L\fI SEV
Set the log severity (see log_levels.h)
.TP
\fB\-s\fR
Show the log severity in the logs
.TP
\fB\-\-decode\fI FLAG
Print \fIFLAG\fR as text and exit

.SH EXAMPLES
.TP
Read a D file from stdin, output to stdout.
cat foo.d | uncrustify \-q \-c my.cfg \-l d
.TP
Process a file, output to stdout.
uncrustify \-c my.cfg \-d foo.d
.TP
Process a source tree, output to a different tree.
find src \-name "*.[ch]" > files.txt
.br
uncrustify \-c my.cfg \-F files.txt \-\-prefix out
.TP
Process a source tree in\(hyplace.
uncrustify \-c my.cfg \-\-no\-backup $(find src \-name "*.[ch]")

.SH NOTES
Use comments containing ' *INDENT\-OFF*' and ' *INDENT\-ON*' to disable
processing of parts of the source file.

.SH AUTHOR
Written by Ben Gardner

.SH REPORTING BUGS
Use the issue tracker at <http://sourceforge.net/projects/uncrustify>

.SH COPYRIGHT
Copyright (C) 2006\(hy2009 Ben Gardner

.SH LICENSE
GNU GPL version 2 or later <http://gnu.org/licenses/gpl.html>

This is free software; see the source for copying conditions.
There is NO warranty; not even for MERCHANTABILITY or FITNESS
FOR A PARTICULAR PURPOSE.



NEWS
No news is good news!



src/uncrustify.c
/**
 * This file takes an input C/C++/D/Java file and reformats it.
 *
 * $Id$
 */
#define DEFINE_GLOBAL_DATA

#include "cparse_types.h"
#include "char_table.h"
#include "chunk_list.h"
#include "prototypes.h"
#include "token_names.h"
#include "args.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <strings.h> /* strcasecmp() */
#include <errno.h>
#include <ctype.h>
#include <sys/stat.h>


static int language_from_tag(const char *tag);
static int language_from_filename(const char *filename);
static const char *language_to_string(int lang);


static void usage_exit(const char *msg, const char *argv0, int code)
{
   if (msg != NULL)
   {
      fprintf(stderr, "%s\n", msg);
   }
   fprintf(stderr,
           "Usage:\n"
           "%s -c cfg [-f file] [-p parsed] [-t typefile] [--version] [-l lang] [-L sev] [-s]\n"
           " c : specify the config file\n"
           " f : specify the file to format, if omitted, stdin is used\n"
           " p : debug - dump parsed tokens to this file\n"
           " L : debug log severities 0-255 for everything\n"
           " s : show log severities\n"
           " t : load a file with types\n"
           " l : language override: C, CPP, D, CS, JAVA\n"
           "--version : print the version and exit\n"
           "The output is dumped to stdout, errors are dumped to stderr\n",
           argv0);
   exit(code);
}

static void version_exit(void)
{
   printf("%s %s\n", PACKAGE, VERSION);
   exit(0);
}

int main(int argc, char *argv[])
{
   struct stat my_stat;
   char        *data;
   int         data_len;
   FILE        *p_file;
   const char  *cfg_file    = "uncrustify.cfg";
   const char  *parsed_file = NULL;
   const char  *source_file = NULL;
   log_mask_t  mask;
   int         idx;
   const char  *p_arg;


   if (argc < 2)
   {
      usage_exit(NULL, argv[0], 0);
   }

   arg_init(argc, argv);

   if (arg_present("--version") || arg_present("-v"))
   {
      version_exit();
   }
   if (arg_present("--help") || arg_present("-h") ||
       arg_present("--usage") || arg_present("-?"))
   {
      usage_exit(NULL, argv[0], 0);
   }

   /* Init logging */
   log_init(stderr);
   if (((p_arg = arg_param("-L")) != NULL) ||
       ((p_arg = arg_param("--log")) != NULL))
   {
      logmask_from_string(p_arg, &mask);
      log_set_mask(&mask);
   }

   /* Get the source file name */
   if (((source_file = arg_param("--file")) == NULL) &&
       ((source_file = arg_param("-f")) == NULL))
   {
      // using stdin
      //usage_exit("Specify the file to process: -f file", argv[0], 57);
   }

   /* Get the config file name */
   if (((cfg_file = arg_param("--config")) == NULL) &&
       ((cfg_file = arg_param("-c")) == NULL))
   {
      usage_exit("Specify the config file: -c file", argv[0], 58);
   }

   /* Get the parsed file name */
   if (((parsed_file = arg_param("--parsed")) != NULL) ||
       ((parsed_file = arg_param("-p")) != NULL))
   {
      LOG_FMT(LNOTE, "Will export parsed data to: %s\n", parsed_file);
   }

   /* Enable log sevs? */
   if (arg_present("-s") || arg_present("--show"))
   {
      log_show_sev(TRUE);
   }

   /* Load type files */
   idx = 0;
   while ((p_arg = arg_params("-t", &idx)) != NULL)
   {
      load_keyword_file(p_arg);
   }

   /* Check for a language override */
   if ((p_arg = arg_param("-l")) != NULL)
   {
      cpd.lang_flags = language_from_tag(p_arg);
      if (cpd.lang_flags == 0)
      {
         fprintf(stderr, "Ignoring unknown language: %s\n", p_arg);
      }
   }

   /* Prepare to load chunks */
   chunk_list_init();


   /* Check for unused args (ignore them) */
   idx = 1;
   while ((p_arg = arg_unused(&idx)) != NULL)
   {
      LOG_FMT(LWARN, "Unused argument: %s\n", p_arg);
   }

   /* Load the config file */
   set_option_defaults();
   if (load_option_file(cfg_file) < 0)
   {
      usage_exit(NULL, argv[0], 56);
   }

   if (source_file == NULL)
   {
      UINT32 data_size;
      int   len;

      if (cpd.lang_flags == 0)
      {
         cpd.lang_flags = LANG_C;
      }

      /* Start with 64k */
      data_size = 64 * 1024;
      data = malloc(data_size);
      data_len = 0;

      while ((len = fread(&data[data_len], 1, data_size - data_len, stdin)) > 0)
      {
         data_len += len;
         if (data_len == data_size)
         {
            data_size += 64 * 1024;
            data = realloc(data, data_size);
            if (data == NULL)
            {
               LOG_FMT(LERR, "Out of memory\n");
               return 100;
            }
         }
      }

      data[data_len] = 0;

      /* Done reading from stdin */
      LOG_FMT(LSYS, "Parsing: %d bytes from stdin as language %s\n",
              data_len, language_to_string(cpd.lang_flags));
   }
   else
   {
      /* Do some simple language detection based on the filename */
      if (cpd.lang_flags == 0)
      {
         cpd.lang_flags = language_from_filename(source_file);
      }

      /* Try to read in the source file */
      p_file = fopen(source_file, "r");
      if (p_file == NULL)
      {
         LOG_FMT(LERR, "open(%s) failed: %s\n", source_file, strerror(errno));
         return(1);
      }

      /*note: could have also just MMAP'd the file */
      fstat(fileno(p_file), &my_stat);

      data_len = my_stat.st_size;
      data     = malloc(data_len + 1);
      fread(data, data_len, 1, p_file);
      data[data_len] = 0;
      fclose(p_file);

      LOG_FMT(LSYS, "Parsing: %s as language %s\n",
              source_file, language_to_string(cpd.lang_flags));
   }

   /**
    * Parse the text into chunks
    */
   tokenize(data, data_len);

   /**
    * Change certain token types based on simple sequence.
    * Example: change '[' + ']' to '[]'
    * Note that level info is not yet available, so it is OK to do all
    * processing that doesn't need to know level info. (that's very little!)
    */
   tokenize_cleanup();

   /**
    * Detect the brace and paren levels and insert virtual braces.
    * This handles all that nasty preprocessor stuff
    */
   brace_cleanup();

   /**
    * At this point, the level information is available and accurate.
    */

   /**
    * Re-type chunks, combine chunks
    */
   fix_symbols();

   /**
    * Look at all colons ':' and mark labels, :? sequences, etc.
    */
   combine_labels();

   /**
    * Change virtual braces into real braces...
    */
   do_braces();

   /**
    * Insert line breaks as needed
    */
   newlines_cleanup_braces();
   if (cpd.settings[UO_nl_squeeze_ifdef])
   {
      newlines_squeeze_ifdef();
   }

   /**
    * Fix same-line inter-chunk spacing
    */
   space_text();

   mark_comments();

   /**
    * Do any aligning of preprocessors
    */
   if (cpd.settings[UO_align_pp_define_span] > 0)
   {
      align_preprocessor();
   }

   /**
    * Indent the text
    */
   indent_text();

   /**
    * Aligning everything else and reindent
    */
   align_all();
   indent_text();

   /**
    * And finally, align the backslash newline stuff
    */
   align_right_comments();
   if (cpd.settings[UO_align_nl_cont] != 0)
   {
      align_backslash_newline();
   }

   if (parsed_file != NULL)
   {
      p_file = fopen(parsed_file, "w");
      if (p_file != NULL)
      {
         output_parsed(p_file);
         fclose(p_file);
      }
   }

   /**
    * Now render it all to the output file
    */
   output_text(stdout);

   return(0);
}


const char *get_token_name(c_token_t token)
{
   if ((token >= 0) && (token < ARRAY_SIZE(token_names)) &&
       (token_names[token] != NULL))
   {
      return(token_names[token]);
   }
   return("???");
}

static BOOL ends_with(const char *filename, const char *tag)
{
   int len1 = strlen(filename);
   int len2 = strlen(tag);

   if ((len2 <= len1) && (strcmp(&filename[len1 - len2], tag) == 0))
   {
      return(TRUE);
   }
   return(FALSE);
}

struct file_lang
{
   const char *ext;
   const char *tag;
   int        lang;
};
struct file_lang languages[] =
{
   { ".c",    "C",    LANG_C },
   { ".h",    "",     LANG_C },
   { ".cpp",  "CPP",  LANG_CPP },
   { ".d",    "D",    LANG_D },
   { ".cs",   "CS",   LANG_CS },
   { ".java", "JAVA", LANG_JAVA },
};

/**
 * Find the language for the file extension
 * Default to C
 *
 * @param filename   The name of the file
 * @return           LANG_xxx
 */
static int language_from_filename(const char *filename)
{
   int i;

   for (i = 0; i < ARRAY_SIZE(languages); i++)
   {
      if (ends_with(filename, languages[i].ext))
      {
         return(languages[i].lang);
      }
   }
   return(LANG_C);
}

/**
 * Find the language for the file extension
 * Default to C
 *
 * @param filename   The name of the file
 * @return           LANG_xxx
 */
static int language_from_tag(const char *tag)
{
   int i;

   for (i = 0; i < ARRAY_SIZE(languages); i++)
   {
      if (strcasecmp(tag, languages[i].tag) == 0)
      {
         return(languages[i].lang);
      }
   }
   return(0);
}

/**
 * Gets the tag text for a language
 *
 * @param lang    The LANG_xxx enum
 * @return        A string
 */
static const char *language_to_string(int lang)
{
   int i;

   for (i = 0; i < ARRAY_SIZE(languages); i++)
   {
      if ((languages[i].lang & lang) != 0)
      {
         return(languages[i].tag);
      }
   }
   return("???");
}

